### PR TITLE
Optimizing query for tasks.update_storage_usage_cache [ENG-1046]

### PR DESCRIPTION
**Pull request by corbinSanders**
## Purpose

Replacing the naive query with a SQL query to optimize tasks.update_storage_usage_cache, as this is run frequently and is inefficient.

## Changes

- api/caching/tasks.py - Modifying update_storage_usage_cache to use a more efficient query.

- I tested this by calling update_storage_usage_cache directly in the shell then making sure the set amount was correct.

## QA Notes

-Ensure this is calculating file size correctly, specifically after moving/copying files.

## Documentation

N/A

## Side Effects

Should not be any

## Ticket

https://openscience.atlassian.net/browse/ENG-1046

### CODE REVIEW

File: api/caching/tasks.py [hunk 1]
```diff
@@ -104,11 +104,20 @@ def ban_url(instance):
...
-    AbstractNode = apps.get_model('osf.AbstractNode')
-
-    storage_usage_total = AbstractNode.objects.get(
-        guids___id=target_id,
-    ).files.aggregate(sum=models.Sum('versions__size'))['sum'] or 0
+    sql = """
+                SELECT sum(version.size) FROM osf_basefileversionsthrough AS obfnv
+                LEFT JOIN osf_basefilenode file ON obfnv.basefilenode_id = file.id
+                LEFT JOIN osf_fileversion version ON obfnv.fileversion_id = version.id
+                LEFT JOIN osf_guid guid ON file.target_object_id = guid.object_id
```
Comment (Johnetordoff) [inline api/caching/tasks.py]: I think you can skip this line if you make sure the target [here](https://github.com/CenterForOpenScience/osf.io/pull/9166/files#diff-a74ffc85f1b853e07331e93ebf2864e3L121) uses `.id` for the pk instead of the guid. You should only have to change one test as well.

File: api/caching/tasks.py [hunk 2]
```diff
@@ -104,11 +104,20 @@ def ban_url(instance):
...
+                WHERE file.provider = 'osfstorage' AND file.target_content_type_id = 4
```
Comment (Johnetordoff) [inline api/caching/tasks.py]: We should get a little comment here that details what content type == 4 or make it explicit that the target is nodes some other way.

File: api/caching/tasks.py (hunk 1 above)
Comment (corbinSanders) [reply api/caching/tasks.py]: Nice catch! One less join.

File: api/caching/tasks.py (hunk 2 above)
Comment (corbinSanders) [reply api/caching/tasks.py]: Ended up joining with django_content_type. Made more sense this way.

**Comment by Johnetordoff** [comment]
💯

File: api/caching/tasks.py [hunk 3]
```diff
@@ -103,19 +103,30 @@ def ban_url(instance):
...
+    sql = """
+        SELECT sum(version.size) FROM osf_basefileversionsthrough AS obfnv
+        LEFT JOIN osf_basefilenode file ON obfnv.basefilenode_id = file.id
+        LEFT JOIN osf_fileversion version ON obfnv.fileversion_id = version.id
+        LEFT JOIN django_content_type type on file.target_content_type_id = type.id
+        WHERE file.provider = 'osfstorage'
+        AND type.model = 'abstractnode'
+        AND file.deleted_on IS NULL
+        AND file.target_object_id=%s
+    """
```
Comment (brianjgeiger) [inline api/caching/tasks.py]: Would it make sense to add a row limit here to minimize memory usage? Maybe something like 5000 rows, and then loop through adding the sums from each query until you've gotten all the rows. It would take a bit longer, but since it wouldn't have to re-plan the query, I doubt it would take too much longer. Plus it's a post-commit task, so it shouldn't hold up any web requests. Could be either a constant defined in the file or an optional parameter for the function with a default.

File: api/caching/tasks.py [hunk 4]
```diff
@@ -103,23 +103,30 @@ def ban_url(instance):
...
-    with connection.cursor() as cursor:
-        cursor.execute(sql, [target_id])
-        result = cursor.fetchall()
-
-    storage_usage_total = int(result[0][0]) if result[0][0] else 0
+    count = per_page
+    offset = 0
+    storage_usage_total = 0
+    while count:
+        with connection.cursor() as cursor:
```
Comment (Johnetordoff) [inline api/caching/tasks.py]: Drive-by: I think you're gonna wanna put this outside the loop.

**Comment by corbinSanders** [comment]
> This looks great. I ran the old query through (with changes for the new through table), and it looks to me like an order of magnitude improvement.
> 
> Old query:
> old query
> 
> New query:
> new query

🎊
